### PR TITLE
Add UseIfEmptyOrIfBlank rule

### LINKED ISSUE #2840: Add a rule to flag places where `ifBlank` and `ifEmpty` can be used
Kotlin provides the functions `ifBlank` and `ifEmpty` so we should flag the points where this functions could be used. Example:

`if (api.name.isBlank()) "John" else api.name`
can be replaced to:
`api.name.ifBlank { "John" }`

The same for `ifEmpty`.

## Context
https://twitter.com/kotlin/status/1277897234043097090

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #2840

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                    return if (list.isEmpty()) {
+                        listOf(1)
+                    } else {
+                        println()
+                        list
+                    }
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).isEmpty()
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt:331]: Why is this not flagged but this is? https://github.com/detekt/detekt/pull/3186/files#diff-1b2a5c91a839b0d25b1cbeae6f734f210415ece8cd0834e477fa5724e1a87c6fR204-R218

**Review by BraisGabin — APPROVED** [review]
Nice!

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import org.jetbrains.kotlin.psi.KtIfExpression
+import org.jetbrains.kotlin.psi.KtThisExpression
+import org.jetbrains.kotlin.psi.psiUtil.blockExpressionsOrSingle
+import org.jetbrains.kotlin.psi.psiUtil.getPossiblyQualifiedCallExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+
+/**
+ * This rule inspects 'isEmpty' or 'isBlank' calls to assign default value. They are can be replaced with 'ifEmpty' or
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:25]:
```suggestion
 * This rule inspects 'isEmpty' or 'isBlank' calls to assign default value. They can be replaced with 'ifEmpty' or
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+                    val name = if (api.name.isEmpty()) "John" else api.name
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings).hasSourceLocation(4, 29)
+            assertThat(findings[0]).hasMessage("This 'isEmpty' call can be replaced with 'ifEmpty'")
+        }
+
+        it("String..isNotEmpty") {
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt:61]:
```suggestion
        it("String.isNotEmpty") {
```

**Review by cortinico — APPROVED** [review]
Great addition 🚀

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt (hunk 1 above)
Comment (t-kameyama) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt:331]: Let's consider following cases:
```kotlin
fun test(list: List<Int>) {
    // case1
    val x = if (list.isEmpty()) {
        println("isEmpty")
        listOf(1)
    } else {
        list
    }

    // case2
    val y = if (list.isEmpty()) {
        listOf(1)
    } else {
        println("isNotEmpty")
        list
    }
}
```

The case1 can be replaced with `ifEmpty`.
```kotlin
fun test(list: List<Int>) {
    // case1
    val x = list.ifEmpty {
        println("isEmpty")
        listOf(1)
    }
}
```

The case2 can't be replaced with `ifEmpty`. There is no place to write `println("isNotEmpty")`.
```kotlin
fun test(list: List<Int>) {
    // case2
    val y = list.ifEmpty {
        listOf(1)
    }
    // println("isNotEmpty")
}
```

**Review by cortinico — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import org.jetbrains.kotlin.psi.KtIfExpression
+import org.jetbrains.kotlin.psi.KtThisExpression
+import org.jetbrains.kotlin.psi.psiUtil.blockExpressionsOrSingle
+import org.jetbrains.kotlin.psi.psiUtil.getPossiblyQualifiedCallExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+
+/**
+ * This rule inspects 'isEmpty' or 'isBlank' calls to assign default value. They can be replaced with 'ifEmpty' or
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:25]:
```suggestion
 * This rule inspects `isEmpty` or `isBlank` calls to assign a default value. They can be replaced with 'ifEmpty' or
```

File: docs/pages/documentation/style.md [hunk 5]
```diff
@@ -1686,6 +1686,39 @@ emptySequence()
 emptySet()
 ```
 
+### UseIfEmptyOrIfBlank
+
+This rule inspects 'isEmpty' or 'isBlank' calls to assign default value. They can be replaced with 'ifEmpty' or
```
Comment (schalkms) [inline docs/pages/documentation/style.md:1691]:
```suggestion
This rule inspects `isEmpty` or `isBlank` calls to assign a default value. They can be replaced with 'ifEmpty' or
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt (hunk 4 above)
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:25]: On second thought I'm not sure, if `isEmpty` and `isBlank` is only used for assigning a default value. Maybe we can come up with a better description.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+                        println()
+                        list
+                    }
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).isEmpty()
+        }
+
+        it("condition is prefix expression") {
```
Comment (schalkms) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt:334]: Why is this condition with a negate operator not reported? I'm not sure whether I get it correctly.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        if (conditionCalleeExpressionText !in conditionFunctionShortNames) return
+
+        val replacement = conditionCallExpression.replacement() ?: return
+        val selfBranch = if (replacement.negativeCondition) thenExpression else elseExpression
+        val selfValueExpression = selfBranch.blockExpressionsOrSingle().singleOrNull() ?: return
+        if (condition is KtDotQualifiedExpression) {
+            if (selfValueExpression.text != condition.receiverExpression.text) return
+        } else {
+            if (selfValueExpression !is KtThisExpression) return
+        }
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:79]:
```suggestion
        } else if (selfValueExpression !is KtThisExpression) {
            return
        }
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    private fun KtExpression.isElseIf(): Boolean = parent.node.elementType == KtNodeTypes.ELSE
+
+    @Suppress("ReturnCount")
+    private fun KtCallExpression.replacement(): Replacement? {
+        val descriptor = getResolvedCall(bindingContext)?.resultingDescriptor ?: return null
+        val receiverParameter = descriptor.dispatchReceiverParameter ?: descriptor.extensionReceiverParameter
+        val receiverType = receiverParameter?.type ?: return null
+        if (KotlinBuiltIns.isArrayOrPrimitiveArray(receiverType)) return null
+        val conditionCallFqName = descriptor.fqNameOrNull() ?: return null
+        return replacements[conditionCallFqName]
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:95]: What happens if `conditionCallFqName` is not in `replacements`?

**Review by schalkms — COMMENTED** [review]
Great PR! Thank you very much! 🙂
Please allow me to have a few questions. Please see my attached comments.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt (hunk 4 above)
Comment (t-kameyama) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt]: 898739c

File: docs/pages/documentation/style.md (hunk 5 above)
Comment (t-kameyama) [reply docs/pages/documentation/style.md]: 898739c

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt (hunk 4 above)
Comment (t-kameyama) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt]: 898739c

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt (hunk 6 above)
Comment (t-kameyama) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlankSpec.kt]: 52b48e7

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt (hunk 7 above)
Comment (t-kameyama) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt]: ec1b4aa

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt (hunk 8 above)
Comment (t-kameyama) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseIfEmptyOrIfBlank.kt:107]: 8d70ba9

**Comment by schalkms** [comment]
The **compile-test-snippets** pre-merge check fails. This weird error can be mitigated by using different class or package names for the test-snippets.

**Review by arturbosch — APPROVED** [review]

**Comment by t-kameyama** [comment]
> The **compile-test-snippets** pre-merge check fails. This weird error can be mitigated by using different class or package names for the test-snippets.

Fixed in 9aa30b5eec02e8d02865bbfc1b43ce447921bcf4.

**Review by schalkms — APPROVED** [review]
Thanks

**Pull request merged**